docs/linking.md

```diff
@@ -0,0 +1,218 @@
+
+By default Slang will provide linkage symbols for all global symbols - types or functions, including declarations. In the future this is likely to change with declarations *not* being exported by default, as the declaration is made available just to forward define within the current compilation, not as a mechnism to descibe something to be imported. 
+
+To mark a declaration as having it's definition defined elsewhere outside the current module, the `[__extern]` attribute can be used. For example 
+
+```
+[__extern] Thing makeThing(int a, int b);
+``` 
+
+With types the `[__extern]` attribute can also be used. For the moment though a type declaration that is externed still needs to be a type definition. Therefore to declare a type to be imported from elsewhere we actually have to write
+
+```
+[__extern] struct Thing {};
+```
+
+Which looks a little like a definition of `Thing` but because of the `[__extern]` it actually means that `Thing` is defined elsewhere.  
+
+Without further language and API support, having every module have all of it's symbols include the module name becomes a problem. If we have two modules `module-a` and `module-b` - how do I use a function defined in `module-a` in `module-b`? There needs to be mechansism/s to allow access to the function as part of the other module. Until the language and API features are added to allow such support, the problem is side stepped by allowing the specifying of a module name. If `module-a` and `module-b` use the same module name they will be able to access one anothers symbols during linkage. 
+
+Specifying the module name can be achieve on the command line with the `-module-name` option as in
+
+```
+slangc -module-name somename -no-codegen tests/serialization/extern/module-a.slang -o tests/serialization/extern/module-a.slang-lib
+```
+
+From the API the module name can be specified via 
+
+```
+spSetDefaultModuleName(compileRequest, "somename");
+```
+
+Linkage
+-------
+
+The linkage process used in Slang is purposefully fairly straight forwards. When linkage occurs all items with linkage have a identifying mangled name. That for some mangled names there could be multiple definitions - for example a declaration of a function and a definition of that function. If there are multiple definitions, linking aims to determine the one that is 'best' which is then used. A definition is better than a declaration. And a definition that is specific to the current target is taken as better than some other definition. 
+
+If there are multiple identical definitions, or definitions where it cannot be determed if one is better than another then one is taken at effectively random. This therefore assumes that if there are multiple definitions that they are in effect the same. 
```

Comment:
Note that this design choice is careful and intentional. It allows us to do things like copy the IR for `inline` functions in module `A` into the IR for any module that `import`s `A`, so that they are guaranteed to be available to optimization passes, but not have it be a linker error if we combine multiple modules that depend on `A` (and thus all have copies of the same `inline` functions).